core/server/services/email-analytics/lib/event-processor.js

```diff
@@ -88,6 +88,16 @@ class GhostEventProcessor extends EventProcessor {
                 opened_at: this.db.knex.raw('COALESCE(opened_at, ?)', [moment.utc(event.timestamp).format('YYYY-MM-DD HH:mm:ss')])
             });
 
+        await this.db.knex('members')
+            .where('email', '=', event.recipientEmail)
+            .andWhere(builder => builder
+                .where('last_seen_at', '<', moment.utc(event.timestamp).subtract(24, 'hours').format('YYYY-MM-DD HH:mm:ss'))
```

Comment:
I think we can improve this a little - we want the behaviour to be "updated once each day" rather than "updated once every 24 hours". i.e. if I visit at 2pm on Monday and then again at 1pm on Tuesday my last_seen should be Tuesday. So I think we can change this to something like (untested)

```suggestion
                .where('last_seen_at', '<', moment.utc(event.timestamp).format('YYYY-MM-DD 00:00:00'))
```

(we can treat last_seen as a date rather than a datetime)

Reply:
The problem you've mentioned would happen in the solution you've shared too. It would happen when a member is visiting at 00:01 UTC and then at 23:59 UTC.
Your solution also means there would be two writes in a short period of time for every member visiting around midnight UTC.

I think the best solution would be to use "updated once every 12 hours" if we require this level of precision, but then it wouldn't match the pitch requirements.

I would choose the "every 12 hours" solution, however I lack context on the reason behind the "last_seen only needs to be updated once in a given 24-hour period" sentence of the pitch. What do you think?

Reply:
The specific wording about a 24-hour period in the pitch is misleading - the thing we care about is the _date_, not the specific time. `last_seen_at` could easily have been just a date field:
> we will not need a write every time a user opens lands on a page, **just the first time that day**.

Reply:
I'm afraid moving to storing `last_seen_at` as a date will have negative consequences for the produce experience.

There will be different kinds of issues depending on how we display the date in the frontend, illustrated here by one member visiting a San Francisco publication (UTC-7) at 3pm and 5pm local time.

### Keeping the UTC date

In this example we just truncate the datetime to 00:00:00 UTC, and display it directly to the publication admins.

- member visits at 1am UTC on Feb, 24th
- last_seen_at is set to 2022-02-24
- The local datetime is 5pm Feb, 23th, the display time is 24th Feb, so the publication admins will see a `last_seen_at` in the future

### Displaying the date in local datetime

In this example we first truncate the datetime to 00:00:00 UTC, then convert it to local time and then display it to the publication admins.

- member visits at 11pm UTC on Feb, 24th
- last_seen_at is set to 2022-02-24
- we convert the date to local time 2022-02-24 00:00:00 UTC = 2022-02-23 16:00:00 UTC
- The local datetime is 3pm Feb, 24th, the display date is Feb, 23th, so the publication admins will see a `last_seen_at` on the previous day

### Conclusion

I recommend not to truncate the `last_seen_at` value. Instead I would recommend to switch to "every 12h" pattern as we have a 1 day precision when displaying the data. The tradeoff seems correct as records will be updated at most twice a day.

I'm fine implementing the date truncation if you are ok with the product limitations it will create.

Reply:
Right - but my proposal wasn't to truncate the date that we store or to switch to a date field, it was to truncate the date that we use for comparison.

The difference is that it gives us a regular, consistent 24h window (== once a day) in which we update the member vs a sliding 12h window.

With the same example:
* member visits at 1am UTC on Feb, 24th
* last_seen_at is set to 2022-02-24 01:00:00
* The local datetime is 5pm Feb 23rd, the display time is 23rd Feb (we display in local timezone)

Reply:
Oh from your previous message I understood that you were suggesting to make `last_seen_at` a date, my bad.
> last_seen_at could easily have been just a date field

### comparing with a truncated date

Do you agree that the first problem you've mentioned will still be present when comparing to a truncated date?

> i.e. if I visit at 2pm on Monday and then again at 1pm on Tuesday my last_seen should be Tuesday.

In the above example that you wrote, if your publication is in UTC-10, then the visit at 1pm on Tuesday will be on the same UTC day.

So depending on the timezone of your publication, there will always be such a problem (although it will be bound to specific times while the 24h isn't).

### regular 24h window

Are consistent 24h windows a good thing? It will create more writes right after midnight UTC across all publications.

Reply:
I've chosen to disagree and commit to the UTC-based window. This has been implemented in the other PR first: https://github.com/TryGhost/Members/pull/370/commits/ff46449ad6069b53cef85a8c2f78daed672668ed